lib/src/front_end/comment_writer.dart

```diff
@@ -0,0 +1,277 @@
+    if (comments.isNotEmpty && _needsSpaceAfterComment(token.lexeme)) {
+      writer.space();
+    }
+  }
+
+  /// Takes all of the comment tokens preceding [token] and converts them to
+  /// [SourceComment]s that track their kind and the whitespace between them.
+  ///
+  /// Returns the list of [SourceComments] and the number of newlines between
+  /// the last comment and [token]. If there are no comments, returns an empty
+  /// list and the number of lines between [token] and the preceding token.
+  (List<SourceComment>, int) _convertComments(Token token) {
+    Token? comment = token.precedingComments;
+
+    // TODO(perf): Avoid calculating newlines between tokens unless there are
+    // comments or it's needed to determine whether to insert a blank in a
+    // sequence.
+
+    // TODO(tall): If the token's comments are being moved by a fix, do not
+    // write them here.
+
+    var previousLine = _endLine(token.previous!);
+    var tokenLine = _startLine(token);
+
+    // Edge case: The analyzer includes the "\n" in the script tag's lexeme,
+    // which confuses some of these calculations. We don't want to allow a
+    // blank line between the script tag and a following comment anyway, so
+    // just override the script tag's line.
+    if (token.previous!.type == TokenType.SCRIPT_TAG) previousLine = tokenLine;
+
+    var comments = <SourceComment>[];
+    while (comment != null) {
+      var commentLine = _startLine(comment);
+
+      // Don't preserve newlines at the top of the file.
+      if (comment == token.precedingComments &&
+          token.previous!.type == TokenType.EOF) {
+        previousLine = commentLine;
+      }
+
+      var text = comment.lexeme.trim();
+      var linesBefore = commentLine - previousLine;
+      var flushLeft = _startColumn(comment) == 1;
+
+      if (text.startsWith('///') && !text.startsWith('////')) {
+        // Line doc comments are always indented even if they were flush left.
+        flushLeft = false;
+
+        // Always add a blank line (if possible) before a doc comment block.
+        if (comment == token.precedingComments) linesBefore = 2;
+      }
+
+      var type = CommentType.block;
+      if (text.startsWith('///') && !text.startsWith('////') ||
+          text.startsWith('/**') && text != '/**/') {
+        // TODO(rnystrom): Check that the comment isn't '/**/' because some of
+        // the dart_style tests use that to mean inline block comments. While
+        // refactoring the Chunk representation to move splits to the front of
+        // Chunk, I want to preserve the current test behavior. The fact that
+        // those tests pass with the old representation is a buggy quirk of the
+        // comment handling.
+        type = CommentType.doc;
+      } else if (comment.type == TokenType.SINGLE_LINE_COMMENT) {
+        type = CommentType.line;
+      } else if (commentLine == previousLine || commentLine == tokenLine) {
+        type = CommentType.inlineBlock;
+      }
```

Comment:
[optional] I missed this defaulting on my first read-through. WDYT about making the default a trailing `else`?

```suggestion
      CommentType type;
      if (text.startsWith('///') && !text.startsWith('////') ||
          text.startsWith('/**') && text != '/**/') {
        // TODO(rnystrom): Check that the comment isn't '/**/' because some of
        // the dart_style tests use that to mean inline block comments. While
        // refactoring the Chunk representation to move splits to the front of
        // Chunk, I want to preserve the current test behavior. The fact that
        // those tests pass with the old representation is a buggy quirk of the
        // comment handling.
        type = CommentType.doc;
      } else if (comment.type == TokenType.SINGLE_LINE_COMMENT) {
        type = CommentType.line;
      } else if (commentLine == previousLine || commentLine == tokenLine) {
        type = CommentType.inlineBlock;
      } else {
        type = CommentType.block
      }
```

Reply:
Good idea. Done in my commit.